File: problems/206/solution2.cpp
// Accepted: 7ms (36.37%), 8.74MB (29.19%)

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    ListNode* reverse(ListNode* node, ListNode* prev = nullptr) {
        if (node == nullptr) {
            return prev;
        }

        ListNode* oldNext = node->next;

        node->next = prev; // reverse

        // return the head of the reversed list
        return reverse(oldNext, node);
    }
public:
    ListNode* reverseList(ListNode* head) {
        return reverse(head);
    }
};

File: problems/206/solution2.ts
// Accepted: 52ms (90.61%), 45.70MB (7.22%)

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function reverseList(head: ListNode | null): ListNode | null {
    const reverse = (node: ListNode | null, prev: ListNode | null = null) => {
        if (node === null) {
            return prev;
        }

        const oldNext = node.next;

        node.next = prev; // reverse

        // return the head of the reversed list
        return reverse(oldNext, node);
    }

    return reverse(head);
};
